File: fraiseql-php/tests/AuthorizationTest.php
<?php

declare(strict_types=1);

namespace FraiseQL\Tests;

use PHPUnit\Framework\TestCase;
use FraiseQL\SchemaRegistry;
use FraiseQL\Security\AuthorizeBuilder;
use FraiseQL\Attributes\GraphQLType;
use FraiseQL\Attributes\Authorize;

/**
 * Tests for custom authorization rules.
 */
final class AuthorizationTest extends TestCase
{
    protected function tearDown(): void
    {
        SchemaRegistry::getInstance()->clear();
        parent::tearDown();
    }

    public function testAuthorizationRuleBuilder(): void
    {
        $config = AuthorizeBuilder::create()
            ->rule("isOwner(\$context.userId, \$field.ownerId)")
            ->description("Ensures users can only access their own notes")
            ->build();

        $this->assertSame("isOwner(\$context.userId, \$field.ownerId)", $config->rule);
        $this->assertSame("Ensures users can only access their own notes", $config->description);
    }

    public function testAuthorizationWithPolicy(): void
    {
        $config = AuthorizeBuilder::create()
            ->policy("piiAccess")
            ->description("References the piiAccess policy")
            ->build();

        $this->assertSame("piiAccess", $config->policy);
        $this->assertTrue($config->cacheable);
    }

    public function testAuthorizationWithErrorMessage(): void
    {
        $config = AuthorizeBuilder::create()
            ->rule("hasRole(\$context, 'admin')")
            ->errorMessage("Only administrators can access this resource")
            ->build();

        $this->assertSame("Only administrators can access this resource", $config->errorMessage);
    }

    public function testAuthorizationRecursiveApplication(): void
    {
        $config = AuthorizeBuilder::create()
            ->rule("canAccessNested(\$context)")
            ->recursive(true)
            ->description("Recursively applies to nested types")
            ->build();

        $this->assertTrue($config->recursive);
    }

    public function testAuthorizationOperationSpecific(): void
    {
        $config = AuthorizeBuilder::create()
            ->rule("isAdmin(\$context)")
            ->operations("create,delete")
            ->description("Only applies to create and delete operations")
            ->build();

        $this->assertSame("create,delete", $config->operations);
    }

    public function testAuthorizationCacheConfiguration(): void
    {
        $config = AuthorizeBuilder::create()
            ->rule("checkAuthorization(\$context)")
            ->cacheable(true)
            ->cacheDurationSeconds(3600)
            ->build();

        $this->assertTrue($config->cacheable);
        $this->assertSame(3600, $config->cacheDurationSeconds);
    }

    public function testAuthorizationNoCaching(): void
    {
        $config = AuthorizeBuilder::create()
            ->rule("checkSensitiveAuthorization(\$context)")
            ->cacheable(false)
            ->build();

        $this->assertFalse($config->cacheable);
    }

    public function testAuthorizationMultipleRules(): void
    {
        $config1 = AuthorizeBuilder::create()
            ->rule("isOwner(\$context.userId, \$field.ownerId)")
            ->description("Ownership check")
            ->build();

        $config2 = AuthorizeBuilder::create()
            ->rule("hasScope(\$context, 'read:notes')")
            ->description("Scope check")
            ->build();

        $this->assertNotSame($config1->rule, $config2->rule);
    }

    public function testAuthorizationFluentChaining(): void
    {
        $config = AuthorizeBuilder::create()
            ->rule("isOwner(\$context.userId, \$field.ownerId)")
            ->description("Ownership authorization")
            ->errorMessage("You can only access your own notes")
            ->recursive(false)
            ->operations("read,update")
            ->cacheable(true)
            ->cacheDurationSeconds(600)
            ->build();

        $this->assertSame("isOwner(\$context.userId, \$field.ownerId)", $config->rule);
        $this->assertSame("Ownership authorization", $config->description);
        $this->assertSame("You can only access your own notes", $config->errorMessage);
        $this->assertFalse($config->recursive);
        $this->assertSame("read,update", $config->operations);
        $this->assertTrue($config->cacheable);
    }

    public function testAuthorizationAttributeBasic(): void
    {
        #[Authorize(rule: "isOwner(\$context.userId, \$field.ownerId)")]
        class ProtectedNote
        {
        }

        $this->assertTrue(true);
    }

    public function testAuthorizationAttributeWithAllParameters(): void
    {
        #[Authorize(
            rule: "isOwner(\$context.userId, \$field.ownerId)",
            description: "Ownership check",
            errorMessage: "Access denied",
            recursive: true,
            operations: "read",
            cacheable: false,
            cacheDurationSeconds: 0,
        )]
        class FullyConfiguredNote
        {
        }

        $this->assertTrue(true);
    }
}

File: fraiseql-csharp/tests/FraiseQL.Security.Tests/AuthorizationTest.cs
using Xunit;
using FraiseQL.Security;
using System.Collections.Generic;

namespace FraiseQL.Security.Tests
{
    public class AuthorizationTest
    {
        [Fact]
        public void ShouldCreateSimpleAuthorizationRule()
        {
            var config = new AuthorizeBuilder()
                .Rule("isOwner($context.userId, $field.ownerId)")
                .Description("Ownership check")
                .Build();

            Assert.Equal("isOwner($context.userId, $field.ownerId)", config.Rule);
            Assert.Equal("Ownership check", config.Description);
        }

        [Fact]
        public void ShouldCreateAuthorizationWithPolicy()
        {
            var config = new AuthorizeBuilder()
                .Policy("ownerOnly")
                .Description("References named policy")
                .Build();

            Assert.Equal("ownerOnly", config.Policy);
        }

        [Fact]
        public void ShouldSupportFluentChaining()
        {
            var config = new AuthorizeBuilder()
                .Rule("hasPermission($context)")
                .Description("Complex rule")
                .ErrorMessage("Access denied")
                .Recursive(true)
                .Operations("read")
                .Build();

            Assert.Equal("hasPermission($context)", config.Rule);
            Assert.True(config.Recursive);
            Assert.Equal("read", config.Operations);
        }

        [Fact]
        public void ShouldSetCachingConfiguration()
        {
            var config = new AuthorizeBuilder()
                .Rule("checkAccess($context)")
                .Cacheable(true)
                .CacheDurationSeconds(600)
                .Build();

            Assert.True(config.Cacheable);
            Assert.Equal(600, config.CacheDurationSeconds);
        }

        [Fact]
        public void ShouldSetErrorMessage()
        {
            var config = new AuthorizeBuilder()
                .Rule("adminOnly($context)")
                .ErrorMessage("Only administrators can access this")
                .Build();

            Assert.Equal("Only administrators can access this", config.ErrorMessage);
        }

        [Fact]
        public void ShouldSetRecursiveApplication()
        {
            var config = new AuthorizeBuilder()
                .Rule("checkNested($context)")
                .Recursive(true)
                .Description("Applied to nested types")
                .Build();

            Assert.True(config.Recursive);
        }

        [Fact]
        public void ShouldSetOperationSpecificRule()
        {
            var config = new AuthorizeBuilder()
                .Rule("canDelete($context)")
                .Operations("delete")
                .Description("Only applies to delete operations")
                .Build();

            Assert.Equal("delete", config.Operations);
        }

        [Fact]
        public void ShouldConvertToDict()
        {
            var config = new AuthorizeBuilder()
                .Rule("testRule")
                .Description("Test")
                .Build();

            var dict = config.ToDict();

            Assert.NotNull(dict);
            Assert.Equal("testRule", dict["rule"]);
            Assert.Equal("Test", dict["description"]);
        }

        [Fact]
        public void ShouldUseAnnotationAttribute()
        {
            var type = typeof(ProtectedResource);
            var attr = type.GetCustomAttributes(typeof(AuthorizeAttribute), false)
                .FirstOrDefault() as AuthorizeAttribute;

            Assert.NotNull(attr);
            Assert.Equal("isOwner($context.userId, $resource.ownerId)", attr.Rule);
        }

        [Fact]
        public void ShouldCreateMultipleConfigurations()
        {
            var config1 = new AuthorizeBuilder()
                .Rule("rule1")
                .Build();

            var config2 = new AuthorizeBuilder()
                .Rule("rule2")
                .Build();

            Assert.NotEqual(config1.Rule, config2.Rule);
        }

        [Fact]
        public void ShouldReturnDefaultCacheSettings()
        {
            var config = new AuthorizeBuilder()
                .Rule("test")
                .Build();

            Assert.True(config.Cacheable);
            Assert.Equal(300, config.CacheDurationSeconds);
        }
    }

    [Authorize(Rule = "isOwner($context.userId, $resource.ownerId)")]
    public class ProtectedResource
    {
        public int Id { get; set; }
        public string Content { get; set; } = "";
    }
}
